components/script/dom/idbopendbrequest.rs

```diff
@@ -0,0 +1,378 @@
+                    let global = this.global();
+                    let cx = global.get_cx();
+
+                    // Step 8.1
+                    let _ac = enter_realm(&*conn);
+                    rooted!(in(*cx) let mut connection_val = UndefinedValue());
+                    unsafe {
+                        conn.to_jsval(*cx, connection_val.handle_mut());
+                    }
+                    this.idbrequest.set_result(connection_val.handle());
+
+                    // Step 8.2
+                    this.idbrequest.set_transaction(&txn);
+
+                    // Step 8.3
+                    this.idbrequest.set_ready_state_done();
+
+                    let event = IDBVersionChangeEvent::new(
+                        &global,
+                        Atom::from("upgradeneeded"),
+                        EventBubbles::DoesNotBubble,
+                        EventCancelable::NotCancelable,
+                        old_version,
+                        Some(version),
+                    );
+
+                    // Step 8.4
+                    txn.set_active_flag(true);
+                    // Step 8.5
+                    let _did_throw = event.upcast::<Event>().fire(this.upcast());
+                    // FIXME:(rasviitanen) Handle throw (Step 8.5)
+                    // https://www.w3.org/TR/IndexedDB-2/#run-an-upgrade-transaction
+                    // Step 8.6
+                    txn.set_active_flag(false);
+
+                    // Implementation specific: we fire the success on db here
+                    // to make sure the success event occurs after the upgrade event.
+                    txn.wait();
+                    this.dispatch_success(&*conn);
+                }),
+                global.upcast(),
+            )
+            .expect("Could not queue task");
+
+        // Step 9: Starts and waits for the transaction to finish
+        transaction.wait();
+    }
+
+    pub fn set_result(&self, result: HandleValue) {
+        self.idbrequest.set_result(result);
+    }
+
+    pub fn set_error(&self, error: Error) {
+        self.idbrequest.set_error(error);
+    }
+
+    pub fn open_database(&self, name: DOMString, version: Option<u64>) {
+        let global = self.global();
+
+        let (sender, receiver) = ipc::channel(global.time_profiler_chan().clone()).unwrap();
+        let response_listener = OpenRequestListener {
+            open_request: Trusted::new(self),
+        };
+
+        let open_operation = SyncOperation::OpenDatabase(
+            sender,
+            global.origin().immutable().clone(),
+            name.to_string(),
+            version,
+        );
+
+        let task_source = global.database_access_task_source();
+
+        let trusted_request = Trusted::new(self);
+        let name = name.to_string();
+        ROUTER.add_route(
+            receiver.to_opaque(),
+            Box::new(move |message| {
+                let global = trusted_request.root().global();
```

Comment:
It is not possible to root DOM objects off of the script thread. Sinec this is only used for passing as an arugment to TaskSource::queue, you will need to use TaskSource::queue_with_canceller instead, and call GlobalScope::task_canceller when calling GlobalScope::database_access_task_source.

This the cause of a huge number of panics in the most recent try run :)